examples/hhl.py

```diff
@@ -0,0 +1,145 @@
+"""Demonstrates the algorithm for solving linear systems by Harrow, Hassidim,
+Lloyd (HHL).
+
+The HHL algorithm solves a system of linear equations, specifically equations
+of the form Ax = b, where A is a Hermitian matrix, b is a known vector, and
+x is the unknown vector. To solve on a quantum system, b must be rescaled to
+have magnitude 1, and the equation becomes:
+
+|x> = A**-1 |b> / || A**-1 |b> ||
+
+The algorithm uses 3 sets of qubits: a single ancilla qubit, a register (to
+store eigenvalues of A), and memory qubits (to store |b> and |x>). The
+following are performed in order:
+1) Quantum phase estimation to extract eigenvalues of A
+2) Controlled rotations of ancilla qubit
+3) Uncomputation with inverse quantum phase estimation
+
+For details about the algorithm, please refer to the paper in the
+REFERENCE section below. The following description uses variables defined
+in the paper.
+
+This example is an implementation of the HHL algorithm for the equation where
+A = [ 1.5  0.5 ]
+    [ 0.5  1.5 ]
+|b> = [ 1  0 ]^T = |0>
+|x> = [ 0.948683  -0.316228 ]
+Eigenvalues of A are 1 (with eigenvector |->) and 2 (with eigenvector |+>)
+
+Because there are 2 eigenvalues, exactly 1 register qubit is needed.
+As a result, t is set to π, and
+
+exp(iAt) = exp(i(H [ 2  0 ] H)*0.5) = H [ exp(i2π)  0 ] H = HZH = X
+                   [ 0  1 ]             [ 0   exp(iπ) ]
+
+|0> for the register qubit corresponds to λ=1, and |1> corresponds to λ=2.
+By setting C = 1, the rotations are Ry(π) for λ=1 and for Ry(π/3) λ=2
+
+Instead of outputting the result directly, this example measures various
+observables of the result.
+
+=== REFERENCE ===
+Coles, Eidenbenz et al. Quantum Algorithm Implementations for Beginners
+https://arxiv.org/pdf/1804.03719.pdf
+
+=== CIRCUIT ===
+(0, 0): ───────────────Ry(π)───────Ry(0.333π)───────────────────M───
+                       │           │                            │
+(1, 0): ───H───@───H───@───────X───@────────────X───H───@───H───┼───
+               │                                        │       │
+(2, 0): ───────X────────────────────────────────────────X───────M───
+
+"""
+
+import cirq
+import math
+
+
+def bitstring(bits):
```

Comment:
WDYT about adding type annotations in this file?

Reply:
We decided that examples should run in Python 2 without modifications (https://github.com/quantumlib/Cirq/issues/755); I think that's why the examples have no type annotations.

Reply:
Makes sense. Thanks.